frontend/src/metabase/lib/revisions/utils.js

```diff
@@ -0,0 +1,173 @@
+    return t`modified question's series`;
+  }
+  return seriesAfter.length > seriesBefore.length
+    ? t`added series to a question`
+    : t`removed series from a question`;
+}
+
+function getCollectionChangeDescription(prevCollectionId, collectionId) {
+  return jt`moved this to ${(
+    <EntityLink entityId={collectionId || "root"} entityType="collections" />
+  )}`;
+}
+
+const CHANGE_DESCRIPTIONS = {
+  // Common
+  name: {
+    [CHANGE_TYPE.UPDATE]: (oldName, newName) =>
+      t`renamed this to` + " " + `"${newName}"`,
+  },
+  description: {
+    [CHANGE_TYPE.ADD]: t`added a description`,
+    [CHANGE_TYPE.UPDATE]: t`changed the description`,
+  },
+  archived: {
+    [CHANGE_TYPE.UPDATE]: (wasArchived, isArchived) =>
+      isArchived ? t`archived this` : t`unarchived this`,
+  },
+  collection_id: {
+    [CHANGE_TYPE.ADD]: getCollectionChangeDescription,
+    [CHANGE_TYPE.UPDATE]: getCollectionChangeDescription,
+    [CHANGE_TYPE.REMOVE]: getCollectionChangeDescription,
+  },
+
+  // Questions
+  dataset_query: {
+    [CHANGE_TYPE.ADD]: t`edited the question`,
+    [CHANGE_TYPE.UPDATE]: t`edited the question`,
+  },
+  visualization_settings: {
+    [CHANGE_TYPE.ADD]: t`changed the visualization settings`,
+    [CHANGE_TYPE.UPDATE]: t`changed the visualization settings`,
+    [CHANGE_TYPE.REMOVE]: t`changed the visualization settings`,
+  },
+
+  // Dashboards
+  cards: {
+    [CHANGE_TYPE.ADD]: (_prevCards, _cards) => {
+      const prevCards = getCardsArraySafe(_prevCards);
+      const cards = getCardsArraySafe(_cards);
+      const count = cards.length - prevCards.length;
+      return ngettext(msgid`added a card`, `added ${count} cards`, count);
+    },
+    [CHANGE_TYPE.UPDATE]: (_prevCards, _cards) => {
+      const prevCards = getCardsArraySafe(_prevCards);
+      const cards = getCardsArraySafe(_cards);
+      if (hasSeriesChange(prevCards) || hasSeriesChange(cards)) {
+        return getSeriesChangeDescription(prevCards, cards);
+      }
+      return t`rearranged the cards`;
+    },
+    [CHANGE_TYPE.REMOVE]: (_prevCards, _cards) => {
+      const prevCards = getCardsArraySafe(_prevCards);
+      const cards = getCardsArraySafe(_cards);
+      const count = prevCards.length - cards.length;
+      return ngettext(msgid`removed a card`, `removed ${count} cards`, count);
+    },
+  },
+};
+
+export function hasDiff(revision) {
+  return Boolean(
+    revision.diff && (revision.diff.before || revision.diff.after),
+  );
+}
+
+export function getChangedFields(revision) {
+  if (!hasDiff(revision)) {
+    return [];
+  }
+  const registeredFields = Object.keys(CHANGE_DESCRIPTIONS);
+
+  // There are cases when either 'before' or 'after' states are null
+  // So we need to pick another one
+  const fields = Object.keys(revision.diff.after || revision.diff.before);
+
+  return fields.filter(field => registeredFields.includes(field));
+}
+
+export function getRevisionDescription(revision) {
```

Comment:
so obviously the sentence fragments created in this function will be wrong for languages that are not subject-verb-object or subject-object-verb, but it doesn't really seem worthwhile to fix (since we don't support any currently), unless you can think of something simple